prio/prg.c

```diff
@@ -0,0 +1,166 @@
+/*
+ * Copyright (c) 2018, Henry Corrigan-Gibbs
+ * 
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. 
+ */
+
+#include <mprio.h>
+#include <nss/pk11pub.h>
+#include <strings.h>
+
+#include "prg.h"
+#include "rand.h"
+#include "share.h"
+#include "util.h"
+
+#define AES_BLOCK_SIZE 16
+
+struct prg {
+  PK11SlotInfo *slot;
+  PK11SymKey *key;
+  PK11Context *ctx;
+};
+
+SECStatus 
+PrioPRGSeed_randomize (PrioPRGSeed *key)
+{
+  return rand_bytes ((unsigned char *)key, PRG_SEED_LENGTH);   
+}
+
+PRG 
+PRG_new (const PrioPRGSeed key_in)
+{
+  PRG prg = malloc (sizeof (*prg));
+  if (!prg) return NULL;
+  prg->slot = NULL;
+  prg->key = NULL;
+  prg->ctx = NULL;
+
+  SECStatus rv = SECSuccess;
+  const CK_MECHANISM_TYPE cipher = CKM_AES_CTR;
+
+  P_CHECKA (prg->slot = PK11_GetBestSlot (cipher, NULL));
```

Comment:
I'd suggest using `PK11_GetInternalSlot` unless you want to use a smart-card or something else to perform the operations. Always using the internal slot will keep you out of trouble.